### PR TITLE
[14.0][IMP] contract

### PR DESCRIPTION
**Pull request by rousseldenis**
I will:

- [x] Update cron process to generate invoices for multi company
- [x] Add tests for multi company
- [x] Add a generation type field to avoid code duplication for sale generation See https://github.com/OCA/contract/pull/661

### CODE REVIEW

**Comment by rousseldenis** [comment]
@manuelregidor @MiquelRForgeFlow This is a more correct solution of https://github.com/OCA/contract/pull/755

See commit : https://github.com/OCA/contract/pull/741/commits/f015ee5fdc0790fbd129c641ca73b7f245018c87

**Comment by rousseldenis** [comment]
I will rebase this

**Comment by flotho** [comment]
hi @rousseldenis , any review needed ?

**Comment by rousseldenis** [comment]
@flotho Of course.

As said, this is more correct than https://github.com/OCA/contract/pull/755/files#diff-bd92412e26bfd77a1e9fb18d4a493b62b8a9c1ec64cbf57a807f244f12c198fcR608

As the search is done once and using with_company() instead of ugly 'allowed_company_ids'

File: contract/models/contract.py [hunk 1]
```diff
@@ -595,19 +599,44 @@ def _recurring_create_invoice(self, date_ref=False):
...
-            contracts_to_invoice = self.search(row["__domain"]).with_context(
-                allowed_company_ids=[row["company_id"][0]]
-            )
-            invoices |= contracts_to_invoice._recurring_create_invoice(date_ref)
-        return invoices
+        # Use ids instead of recordset for perfomances reasons
+        for company in companies:
+            contracts_to_invoice = contracts.filtered(
+                lambda c: c.company_id == company
+            ).with_company(company)
```
Comment (rousseldenis) [inline contract/models/contract.py:635]: @MiquelRForgeFlow Could you please review this ?

I want to use with_company that is better than using allowed_company_ids in context (that part is in Odoo core).

**Review by carlosdauden — APPROVED** [review]
Resolve conflicts

**Review by MiquelRForgeFlow — APPROVED** [review]
LGTM

File: contract/models/abstract_contract.py [hunk 2]
```diff
@@ -48,6 +48,20 @@ class ContractAbstractContract(models.AbstractModel):
...
     )
+    generation_type = fields.Selection(
+        string="Generation Type",
+        selection=lambda self: self._get_generation_type_selection(),
+        default=lambda self: self._get_default_generation_type(),
+        help="Choose the document that will be automatically generated by cron.",
+    )
+
+    @api.model
+    def _get_generation_type_selection(self):
```
Comment (pedrobaeza) [inline contract/models/abstract_contract.py]: Please use conventions for method naming (https://www.odoo.com/documentation/15.0/es/developer/misc/other/guidelines.html#symbols-and-conventions): `_selection_generation_type` and `_default_generation_type`.

File: contract/models/contract.py [hunk 3]
```diff
@@ -595,19 +599,44 @@ def _recurring_create_invoice(self, date_ref=False):
...
+            getattr(contracts_to_invoice, _recurring_create_func)(date_ref)
+        return True
+
+    @api.model
+    def cron_recurring_create_invoice(self, date_ref=None):
+        return self._cron_recurring_create(date_ref)
```
Comment (pedrobaeza) [inline contract/models/contract.py]: I would suggest the explicit form for avoiding misunderstandings:

```suggestion
        return self._cron_recurring_create(date_ref, create_type="invoice")
```

File: contract/tests/test_contract.py [hunk 4]
```diff
@@ -2412,3 +2412,7 @@ def test_contract_action_preview(self):
         action = self.contract.action_preview()
         self.assertIn("/my/contracts/", action["url"])
         self.assertIn("access_token=", action["url"])
+
+
+class TestContractCase(TestContract, common.SavepointCase):
```
Comment (pedrobaeza) [inline contract/tests/test_contract.py]: Why adding this?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: contract/models/abstract_contract.py (hunk 2 above)
Comment (rousseldenis) [reply contract/models/abstract_contract.py]: Arf didn't know Odoo has done such requirements :smile: But great!

**Comment by rousseldenis** [comment]
@MiquelRForgeFlow Rebased on your last work
@pedrobaeza Requested changes attended

**Comment by rousseldenis** [comment]
@pedrobaeza This one is ready

File: contract/tests/test_contract.py [hunk 5]
```diff
@@ -17,7 +17,7 @@ def to_date(date):
     return fields.Date.to_date(date)
 
 
-class TestContractBase(common.SavepointCase):
+class TestContractBase:
```
Comment (pedrobaeza) [inline contract/tests/test_contract.py]: Please don't remove the base inheritance from here.

File: contract/models/contract.py [hunk 6]
```diff
@@ -595,23 +599,45 @@ def _recurring_create_invoice(self, date_ref=False):
...
-            )
-            invoices |= contracts_to_invoice._recurring_create_invoice(date_ref)
-        return invoices
+        # Use ids instead of recordset for perfomances reasons
+        for company in companies:
+            contracts_to_invoice = contracts.filtered(
+                lambda c: c.company_id == company
+                and (not c.date_end or c.recurring_next_date <= c.date_end)
+            ).with_company(company)
+            getattr(contracts_to_invoice, _recurring_create_func)(date_ref)
```
Comment (pedrobaeza) [inline contract/models/contract.py]: I don't like too much this obscure way to execute methods. When you try to search for calling occurrences, you don't get anything, and that's very confusing. You can prepare a hook that explicitly call each method depending on the type.

**Review by pedrobaeza — COMMENTED** [review]
Maybe you can squash a bit the commit history: 10 commits for 6 files seems too much.

File: contract/tests/test_contract.py (hunk 5 above)
Comment (rousseldenis) [reply contract/tests/test_contract.py]: That's done with purpose as tests for multi company import account base ones that inherits from SavePointCase.

File: contract/models/contract.py (hunk 6 above)
Comment (rousseldenis) [reply contract/models/contract.py]: Solved

File: contract/tests/test_contract.py (hunk 5 above)
Comment (pedrobaeza) [reply contract/tests/test_contract.py]: But this forces all of the inherited tests to multi-inherit from SavepointCase and this one, being very prompt to be forgotten and not having real tests performed. If you keep it, is there any error?

Comment (rousseldenis) [reply contract/tests/test_contract.py]: Of course, you cannot inherit from same one twice.

**Comment by rousseldenis** [comment]
@pedrobaeza 

> Maybe you can squash a bit the commit history: 10 commits for 6 files seems too much.

No, it reflects development

File: contract/tests/test_contract.py (hunk 5 above)
Comment (pedrobaeza) [reply contract/tests/test_contract.py]: Well, I don't like to change this base structure, due to the reasons above. Can you please find another solution?

**Comment by rousseldenis** [comment]
So this one is really ready :smile: 
@pedrobaeza

File: contract/models/contract.py [hunk 7]
```diff
@@ -595,23 +599,47 @@ def _recurring_create_invoice(self, date_ref=False):
...
-            contracts_to_invoice = (
-                self.search(row["__domain"])
-                .with_context(allowed_company_ids=[row["company_id"][0]])
-                .filtered(
-                    lambda a: not a.date_end or a.recurring_next_date <= a.date_end
-                )
-            )
-            invoices |= contracts_to_invoice._recurring_create_invoice(date_ref)
-        return invoices
+        # Use ids instead of recordset for perfomances reasons
```
Comment (pedrobaeza) [inline contract/models/contract.py]: I think this comment doesn't apply anymore, isn't it?

Comment (rousseldenis) [reply contract/models/contract.py]: Done

**Review by pedrobaeza — APPROVED** [review]
Thanks for the efforts. Do you plan to fw-port it to v15?

**Comment by MiquelRForgeFlow** [comment]
@rousseldenis will you forward port this?

**Comment by rousseldenis** [comment]
> @rousseldenis will you forward port this?

I'll do oca-port yes!

**Comment by rousseldenis** [comment]
> @rousseldenis will you forward port this?

Done: https://github.com/OCA/contract/pull/776

**Pull request merged**